src/rag/reranker/reranker.py

```diff
@@ -0,0 +1,109 @@
+from typing import Protocol
+
+from langchain_core.documents import Document
+from langchain_core.output_parsers import PydanticOutputParser
+from langchain_core.prompts import PromptTemplate
+from pydantic import BaseModel
+
+from rag.reranker.prompt import RERANKER_PROMPT_TEMPLATE
+from rag.reranker.rrf import get_relevant_documents
+from rag.reranker.utils import document_to_str
+from utils.logging import get_logger
+from utils.models.factory import IModel
+
+logger = get_logger(__name__)
+
+
+class RerankedDoc(BaseModel):
+    """A Reranked Document with page content."""
+
+    page_content: str
+
+
+class RerankedDocs(BaseModel):
+    """A list of Reranked Documents."""
+
+    documents: list[RerankedDoc]
+
+
+class IReranker(Protocol):
+    """Interface for RAG rerankers."""
+
+    def rerank(
+        self,
+        docs_list: list[list[Document]],
+        queries: list[str],
+        input_limit: int = 10,
+        output_limit: int = 4,
+    ) -> list[Document]:
+        """Rerank the documents based on which documents are most relevant to the given queries."""
+        ...
+
+
+class LLMReranker(IReranker):
+    """Reranker based on a language model."""
+
+    def __init__(self, model: IModel):
+        """Initialize the reranker."""
+        reranked_docs_parser = PydanticOutputParser(pydantic_object=RerankedDocs)
+        prompt = PromptTemplate.from_template(RERANKER_PROMPT_TEMPLATE).partial(
+            format_instructions=reranked_docs_parser.get_format_instructions()
+        )
+        self.chain = prompt | model.llm | reranked_docs_parser
+
+    def rerank(
+        self,
+        docs_list: list[list[Document]],
+        queries: list[str],
+        input_limit: int = 10,
+        output_limit: int = 4,
+    ) -> list[Document]:
+        """
+        Rerank the documents based on which documents are most relevant to the given queries.
+        The documents are first filtered based on their relevance to the queries and capped at the input limit.
+        Then, the LLM reranker is used to rerank the filtered documents and capped at the output limit.
+        Finally, the reranked documents are returned.
+        :param docs_list: A list of lists of documents.
+        :param queries: A list of queries.
+        :param input_limit: The maximum number of documents to consider for reranking.
+        :param output_limit: The maximum number of documents to return.
+        :return: A list of reranked documents.
+        """
+        logger.info(f"Reranking documents for queries: {queries}")
+
+        # filtration to prevent reranking irrelevant documents
+        relevant_docs = get_relevant_documents(docs_list, limit=input_limit)
+
+        # reranking using the LLM model
+        response: RerankedDocs = self.chain.invoke(
```

Comment:
```suggestion
        response: RerankedDocs = await self.chain.ainvoke(
```

Reply:
please add error handling for reranking LLM call,  so that it fallbacks. It seems sometimes it cannot return the expected format and fails with parsing. This happened for mytesting. For example:
```
        try:
            response: RerankedDocs = await self.chain.ainvoke(
                {
                    "documents": format_documents(relevant_docs),
                    "queries": format_queries(queries),
                    "limit": output_limit,
                }
            )
        except Exception as e:
            logger.error(f"Error during document reranking: {str(e)}")
            # Fallback to using the filtered documents
            return relevant_docs[:output_limit]
```

Reply:
we really need to test prompt of Reranker with separate integration tests. You can create a separate method for calling the chain only. Then it'll be really easy to test your reranking logic separately: 
```
def _anvoke_reranker():
  self.chain.invoke(...)
```

If you say too much for the PR, let's do it in a follow-up PR.